fog/ledger/server/src/key_image_router_service.rs

```diff
@@ -0,0 +1,83 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+use std::{
+    collections::HashMap,
+    sync::{Arc, RwLock},
+};
+
+use futures::{FutureExt, TryFutureExt};
+use grpcio::{DuplexSink, RequestStream, RpcContext};
+use mc_common::logger::{log, Logger};
+use mc_fog_api::{
+    ledger::{LedgerRequest, LedgerResponse},
+    ledger_grpc::{self, LedgerApi},
+};
+use mc_fog_ledger_enclave::LedgerEnclaveProxy;
+use mc_fog_uri::KeyImageStoreUri;
+use mc_util_grpc::rpc_logger;
+use mc_util_metrics::SVC_COUNTERS;
+
+use crate::router_handlers;
+
+#[derive(Clone)]
+pub struct KeyImageRouterService<E>
+where
+    E: LedgerEnclaveProxy,
+{
+    enclave: E,
+    shards: Arc<RwLock<HashMap<KeyImageStoreUri, Arc<ledger_grpc::KeyImageStoreApiClient>>>>,
```

Comment:
Is the inner `Arc<>` around `ledger_grpc::KeyImageStoreApiClient` needed? You probably put this here for a good reason, I'm just a little confused since this global mutex is the only thing that needs to hold onto it, right? Or is there somewhere that accessing it with a borrow doesn't quite fit?

Reply:
Hm, I agree that it seems superfluous, but I think I did have a reason... I'll dig into it to refresh my memory.

Reply:
Ah yes, so the reason for this is that the `Vec<Arc<KeyImageStoreApiClient>>` in `handle_requests` in `router_handlers.rs` gets passed to `handle_request` for every request in the stream, which is each a separate future. So it needs to be cloned into that future, and cloning the whole `Vec` would presumably be expensive, but cloning a `Vec<Arc<>>` is cheaper. I think it could maybe be an `Arc<Vec<>>` instead, and that might be somewhat more "correct", but I suspect it would have some other effect. There might also be a nicer way to do this if grpcio's async backend wasn't so weird.

I also think this will end up changing somewhat when we make the changes related to knowing about shard block ranges.